### PR TITLE
Fixed #15579 - Added ability to skip parent deletion while deleting a child

### PR DESCRIPTION
**Pull request by coldmind**
https://code.djangoproject.com/ticket/15579

This is an example how this can be implemented.
If it is good-to-go, I will add some documentation.

### CODE REVIEW

File: django/db/models/deletion.py [hunk 1]
```diff
@@ -200,20 +203,21 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
...
-                # FIXME: This seems to be buggy and execute a query for each
-                # parent object fetch. We have the parent data in the obj,
-                # but we don't have a nice way to turn that data into parent
-                # object instance.
-                parent_objs = [getattr(obj, ptr.name) for obj in new_objs]
-                self.collect(parent_objs, source=model,
-                             source_attr=ptr.rel.related_name,
-                             collect_related=False,
-                             reverse_dependency=True)
+        if keep_parents:
```
Comment (collinanderson) [inline django/db/models/deletion.py]: if _not_ keep_parents :)

File: tests/delete/tests.py [hunk 2]
```diff
@@ -349,6 +350,22 @@ def test_large_delete_related(self):
         self.assertFalse(S.objects.exists())
         self.assertFalse(T.objects.exists())
 
+    def test_delete_with_no_parents_deletion(self):
+        child = RChild.objects.create()
+        parent_id = child.r_ptr_id
+        child.delete(keep_parents=False)
```
Comment (collinanderson) [inline tests/delete/tests.py]: `keep_parents=True` :)

File: django/db/models/deletion.py (hunk 1 above)
Comment (coldmind) [inline django/db/models/deletion.py]: @collinanderson really :) I did it just before going to sleep

**Comment by coldmind** [comment]
something with git checkout in builds...
buildbot, test this please.

**Comment by coldmind** [comment]
buildbot, test this please.

File: tests/delete/tests.py [hunk 3]
```diff
@@ -349,6 +350,22 @@ def test_large_delete_related(self):
...
         self.assertFalse(T.objects.exists())
 
+    def test_delete_with_no_parents_deletion(self):
+        child = RChild.objects.create()
+        parent_id = child.r_ptr_id
+        child.delete(keep_parents=True)
+        self.assertFalse(RChild.objects.filter(id=child.id).exists())
+        self.assertTrue(R.objects.filter(id=parent_id).exists())
+
+    def test_delete_grandchild_not_affects_parents(self):
```
Comment (MarkusH) [inline tests/delete/tests.py]: I think this test is sufficient, no need for the other one. Can you rename it to `test_delete_keep_parents()` though.

**Comment by MarkusH** [comment]
LGTM, documentation is yet to be done:

* `docs/ref/models/instances.txt`: "Deleting objects"
* `docs/releases/1.9.txt`: "What's new in Django 1.9" --> "Minor features" --> "Models"

**Comment by MarkusH** [comment]
The docs warning is
```
docs/docs/releases/1.9.txt:135: WARNING: py:meth reference target not found: Model.delete
```

File: docs/ref/models/instances.txt [hunk 4]
```diff
@@ -533,7 +533,7 @@ value, the field will be added to the updated fields.
 Deleting objects
 ================
 
-.. method:: Model.delete([using=DEFAULT_DB_ALIAS])
+.. method:: Model.delete([using=DEFAULT_DB_ALIAS, keep_parents=False])
```
Comment (timgraham) [inline docs/ref/models/instances.txt]: should have a `.. versionchanged:: 1.9` annotation at the bottom of this method indicating that the `keep_parents` parameter was added.

**Comment by timgraham** [comment]
merged in 81c2d9f60b9206c1291e5b1c3c8686f24a7726e1, thanks!

**Pull request closed without merging**